docs/source/advanced/profiler.rst

```diff
@@ -191,16 +197,103 @@ The user can provide ``PyTorchProfiler(record_functions={...})`` to extend the s
     Self CPU time total: 1.681ms
 
-When running with `PyTorchProfiler(emit_nvtx=True)`. You should run as following::
+When running with ``PyTorchProfiler(emit_nvtx=True)``, you should run as following:
+
+.. code-block::
 
     nvprof --profile-from-start off -o trace_name.prof -- <regular command here>
 
 To visualize the profiled operation, you can either:
 
-Use::
+.. code-block::
 
     nvvp trace_name.prof
 
-Or::
+.. code-block::
 
     python -c 'import torch; print(torch.autograd.profiler.load_nvprof("trace_name.prof"))'
+
+
+----------------
+
+****************
+Custom Profiling
+****************
+
+Custom Profiler
+===============
+
+You can also configure a custom profiler and pass it to the Trainer. To configure it, subclass :class:`~pytorch_lightning.profiler.base.BaseProfiler`
+and override some of its methods. The following is a simple example that profiles the first occurance and total calls of each action:
+
+.. code-block:: python
+
+    from pytorch_lightning.profiler.base import BaseProfiler
+    from collections import defaultdict
+    import time
+
+
+    class ActionCountProfiler(BaseProfiler):
+        def __init__(self, dirpath=None, filename=None):
+            super().__init__(dirpath=dirpath, filename=filename)
+            self._action_count = defaultdict(int)
+            self._action_first_occurance = {}
+
+        def start(self, action_name):
+            if action_name not in self._action_first_occurance:
+                self._action_first_occurance[action_name] = time.strftime("%m/%d/%Y, %H:%M:%S")
+
+        def stop(self, action_name):
+            self._action_count[action_name] += 1
+
+        def summary(self):
+            res = f"\nProfile Summary: \n"
+            max_len = max(len(x) for x in self._action_count)
+
+            for action_name in self._action_count:
+                # generate summary for actions called more than once
+                if self._action_count[action_name] > 1:
+                    res += (
+                        f"{action_name:<{max_len}s} \t "
+                        + "self._action_first_occurance[action_name]} \t "
+                        + "{self._action_count[action_name]} \n"
+                    )
+
+            return res
+
+        def teardown(self, stage):
+            self._action_count = {}
+            self._action_first_occurance = {}
+            super().teardown(stage=stage)
+
+.. code-block:: python
+
+    trainer = Trainer(..., profiler=ActionCountProfiler())
+    trainer.fit(...)
+
+
+Profile Logic of Your Interest
+==============================
+
+You can also reference this profiler in your LightningModule to profile specific actions of interest.
+Each profiler has a method ``profile()``m which returns a context handler. Simply pass in the name of
+your action that you want to track and the profiler will record performance for code executed within this context.
```

Comment:
Profilers can be used outside LightningModule, too, right?

```suggestion
You can also reference this profiler in your script to profile specific actions of interest.
Each profiler has a method ``profile()``m which returns a context handler. Simply pass in the name of
your action that you want to track and the profiler will record performance for code executed within this context.
```

Reply:
yes.. but these statements are explaining the attached example. You can refer them anywhere anyways as long as teardown is called to clear the cache.